### PR TITLE
Check cout and cerr usage in make style + Improved script messages [make-style-cout-cerr-check]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
Add cout and cerr usage check to "make style" script.
Travis fails if this check is failed, therefore check should be made easier.
code_style script requires changes to be added, and gives a non descriptive error.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1267](https://github.com/mfem/mfem/pull/1267) | @IdoAkkerman | @tzanio | @adrienbernede + @pazner | 02/16/20 | 02/26/20 | 02/27/20 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
I see your point, code_style does require a clean history to perform a relevant diff. And you would like to be able to check before commit happens.

I can’t help regretting the duplication, but I also think you are right about checking before commit.

Now, it seems to me that you one not want to run "make style" before a commit. In fact, it would mix code reformat with the not committed changes. Let’s say for example that one mistakenly uses a different version of astyle, this would result in a major rewriting. Reset to HEAD would result in the loss of the changes introduced.

The conclusion is "make style is not a _check_", and we want a _check_ for std::cout.

**Comment by IdoAkkerman** [comment]
Perhaps instead of a check it should be a replacement:
cout --> mfem::out
std::cout --> mfem::out
Risk here, and with the grep btw, is that it might pick up on spurious "cout"s.

**Comment by IdoAkkerman** [comment]
I see your point, however, a check is less intrusive then a change. 
So adding a check to a change script is not a big deal IMO.
Adding a change script to a check would be different thing.

The scenario you sketch now can also happen without the check.

**Comment by adrienbernede** [comment]
@IdoAkkerman I am referring to your specific needs: launching `code-style` before commit.

My point was: make style is not meant to be launched before commit, so adding the check to this wouldn’t help, because code-style can already handle after commit checks/changes.

**Comment by tzanio** [comment]
I am conflicted if we want to have things like this in the makefile or if they really belong in scripts (which should be better documented).

What do you guys think?

| &nbsp; | YES, include in makefile | NO, keep in scripts |
| ---- | ---- | ---- |
@acfisher  | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@adrienbernede | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@camierjs  | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@jandrej  | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@tzanio  | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@v-dobrev | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>
@vladotomov  | <ul><li>- [ ] </li></ul> | <ul><li>- [ ] </li></ul>

**Comment by v-dobrev** [comment]
> My point was: make style is not meant to be launched before commit ...

This is not quite right -- `make style` is meant to be applied at any point to fix the styling of the code. If we worry that someone may apply changes with the wrong astyle version, we can add a version check to the `style` target.

On the other hand, replacing `cout`/`cerr` is not quite a "style" change but if it can be reliably applied from a script, it is not a big leap to consider it a "style" fix.

We can add a shell script `apply-style` that will be called by `make style` -- if the script itself is not too long, we can just put it in the `style` target. Either way, since the `check-style` script calls `make style`, any changes related to the `cout`/`cerr` will be in one place (and they can be removed from `check-style`).

**Comment by v-dobrev** [comment]
I just looked at the changes in the PR and I see that, as it is, there is no change in behavior for the `make style` target itself (in terms of changed files) -- it just adds a warning-like message. However, we should consider its effect on the `check-style` script.

**Comment by adrienbernede** [comment]
OK for checking astyle version check.

> On the other hand, replacing cout/cerr is not quite a "style" change but if it can be reliably applied from a script, it is not a big leap to consider it a "style" fix.

AFAIK this is not easily feasible.

> code_style script requires changes to be added, and gives a non descriptive error.

What am I missing here? I cannot see why this is the case.

**Comment by adrienbernede** [comment]
I think it is preferable to have all checks/tests as scripts. There could be `style-check` (astyle check + cout cerr check), `apply-astyle` (not ambiguous).
`style-check` could prompt "Do you want to apply astyle (won’t solve std::cout std::cerr issues)?"

I am no found of adding plenty targets to makefile because of the duplication of feature with existing scripts. "You can run either `make style` or `cd tests/scripts && ./runtests code-style` or `cd tests/scripts && code-style ../..` that’s the same...". Especially when we have a second build system (manager) to maintain (cmake . style ?).

Script can better documented, and scripts can be used in githooks, and one could create aliases to ease their use.

If that’s really what you want, then calling scripts in makefile targets is an intermediate solution. Although again, duplicating features.

**Comment by IdoAkkerman** [comment]
I think  script-only would be the prefered way, but perhaps delete `make style` or give deprecate warning and a pointer to `runtest check_style` and `runtest apply_style`.

In both `check_style` and `apply_style` the cout/cerr should only be a warning (perhaps harsher for apply). I don't think a replace using for instance sed would be desirable, due the non-zero chance of false positives (for instance an output counter defined as `int cout`) .

**Comment by IdoAkkerman** [comment]
The cout/cerr warning should be more descriptive, as in `make style`.  
It was quite frustrating to figure it out, waste time, and then realize it is one of those stupid errors that you should have known.

**Comment by v-dobrev** [comment]
I think this is OK as it is. Additional improvements can be made later, if someone has the time and desire to make them.

**Comment by tzanio** [comment]
I took the liberty of updating the makefile and the scripts to hopefully provide more useful information to the users that are new to do. Please let me know what you think.

@pazner, can you turn https://github.com/mfem/mfem/issues/1255#issuecomment-586528331 into a test script that could be called in Travis (or incorporate directly in Travis)? If that works better, we can retire `branch-history`.

**Comment by pazner** [comment]
> @pazner, can you turn [#1255 (comment)](https://github.com/mfem/mfem/issues/1255#issuecomment-586528331) into a test script that could be called in Travis (or incorporate directly in Travis)? If that works better, we can retire `branch-history`.

@tzanio yes, I have improved the script and it can easily be called from Travis to replace `branch-history`. What is the best place to introduce that script (e.g. new PR, this PR, ...)?

**Comment by adrienbernede** [comment]
@tzanio @pazner 
Should this test be callable by run_test ? I remember it was a requirement for branch-history.
Maybe it is already...

**Comment by pazner** [comment]
@adrienbernede the updated version is callable by `runtest`

**Comment by adrienbernede** [comment]
@pazner, not a requirement, but is it safe against `Ctrl-C` interruption? It was on my checklist for an update of `branch-history`.

I think it should be another PR, you can assign me to it.

**Comment by pazner** [comment]
@adrienbernede yes, it is safe with `Ctrl-C` since it does not change the git state.

File: tests/scripts/code-style [hunk 1]
```diff
@@ -57,5 +57,19 @@ make style
...
+# Tell users what to do if the script fails
+if [ -s "$0.err" ]; then
+    cat >  $0.msg <<EOF
+
+To correct this error, run 'make style' and make sure that:
+
+1) The code is formatted with Artistic Style 2.05.1.
+2) There is no use of std::cout/cerr (use mfem::out/err instead).
+
+Please contact a member of the MFEM team if you have questions.
```
Comment (adrienbernede) [inline tests/scripts/code-style]: I would suggest to change the suggestion as follow:

```
To correct this error:
- First make sure that
  1) you are using Artistic Style 2.05.1 ("> astyle --version"),
  2) there is not use of (std::)cout/cerr (use (mfem::)out/err instead).
- Then run "> make style" to apply astyle formatting and verify your code.

Please contact a member of the MFEM team if you have questions.
```

**Comment by tzanio** [comment]
> @tzanio yes, I have improved the script and it can easily be called from Travis to replace `branch-history`. What is the best place to introduce that script (e.g. new PR, this PR, ...)?

It is up to you. Maybe a new PR is indeed cleaner.

**Comment by pazner** [comment]
Ok, I made a new PR: #1309 (WIP for now)

**Comment by tzanio** [comment]
@pazner, can you take a quick look? I made a lot of the recent changes, so I probably shouldn't be the second reviewer.

**Comment by tzanio** [comment]
@pazner, can you take a quick look?

File: tests/scripts/gitignore [hunk 2]
```diff
@@ -55,5 +55,28 @@ EOF
...
 
+# Tell users what to do if the script fails
+if [ -s "$0.err" ]; then
+    cat >  $0.msg <<EOF
+
+To correct this error, make sure that the .gitignore file is properly updated.
+
+Generally, this file should contain exactly the files that can be generated from
+the repository (via building + running) that we *do not* want in the repository.
+(This was such files cannot be accidentally committed.)
```
Comment (pazner) [inline tests/scripts/gitignore]:
```suggestion
(This way such files cannot be accidentally committed.)
```

File: tests/scripts/runtest [hunk 3]
```diff
@@ -99,10 +99,11 @@ do
    if [ -s "$this_test.err" ]; then
       runtest_check=1
       printf '\033[31m'"FAILED\n"'\033[0m'
+      (printf '\033[31m'; cat $this_test.msg; printf '\033[0m') 1>&2
       echo "=== BEGIN error log from $this_test.err"
       (cat $this_test.err) 1>&2
       echo "=== END error log from $this_test.err"
-      # rm -f $this_test.err $this_test.out
+      # rm -f $this_test.err $this_test.out $this_test.msg
```
Comment (pazner) [inline tests/scripts/runtest]: Why is this commented out?

Comment (tzanio) [reply tests/scripts/runtest]: These could be useful for debugging and maybe more, but I'll go ahead and delete them for now... folks can always comment that line if necessary.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**